fish-rust/src/job_group.rs

```diff
@@ -0,0 +1,352 @@
+use self::job_group::pgid_t;
+use crate::common::{assert_send, assert_sync};
+use crate::wchar_ffi::{WCharFromFFI, WCharToFFI};
+use cxx::{CxxWString, UniquePtr};
+use std::num::{NonZeroI32, NonZeroU32};
+use std::sync::atomic::{AtomicBool, AtomicI32, Ordering};
+use std::sync::Mutex;
+use widestring::WideUtfString;
+
+#[cxx::bridge]
+mod job_group {
+    // Not only does cxx bridge not recognize libc::pid_t, it doesn't even recognize i32 as a POD
+    // type! :sadface:
+    struct pgid_t {
+        value: i32,
+    }
+
+    extern "Rust" {
+        #[cxx_name = "job_group_t"]
+        type JobGroup;
+
+        fn wants_job_control(&self) -> bool;
+        fn wants_terminal(&self) -> bool;
+        fn is_foreground(&self) -> bool;
+        fn set_is_foreground(&self, value: bool);
+        #[cxx_name = "get_command"]
+        fn get_command_ffi(&self) -> UniquePtr<CxxWString>;
+        #[cxx_name = "get_job_id"]
+        fn get_job_id_ffi(&self) -> i32;
+        #[cxx_name = "get_cancel_signal"]
+        fn get_cancel_signal_ffi(&self) -> i32;
+        #[cxx_name = "cancel_with_signal"]
+        fn cancel_with_signal_ffi(&self, signal: i32);
+        fn set_pgid(&mut self, pgid: i32);
+        #[cxx_name = "get_pgid"]
+        fn get_pgid_ffi(&self) -> UniquePtr<pgid_t>;
+        fn has_job_id(&self) -> bool;
+
+        // cxx bridge doesn't recognize `libc::*` as being POD types, so it won't let us use them in
+        // a SharedPtr/UniquePtr/Box and won't let us pass/return them by value/reference, either.
+        unsafe fn get_modes_ffi(&self, size: usize) -> *const u8; /* actually `* const libc::termios` */
```

Comment:
Interesting. Another option is to define an opaque Rust type like we do for

	type Features;
	fn mutable_fish_features() -> *mut Features;